Small improvements to Matrix mobject

PhilippImhof opened the pull request:
easy customization of the left and right bracket of a matrix
- added attributes `left_bracket` and `right_bracket` to `Matrix` class, default is to use old hard-coded brackets
- changed `add_brackets()` to accept parameters specifying left and right bracket to use
- changed `__init__()` in order to call `add_brackets` with the newly introduced parameters

set row colors to a matrix mobject, similar function exists for columns
- added method `get_rows()` similar to `get_columns()`, returning the rows of a matrix as a `VGroup`
- added method `set_row_colors()` similiar to `set_column_colors()`

PhilippImhof commented:
This PR has been sent to 3b1b's repository a few months ago, but never got merged as (probably) Grant did not consider this useful to his own projects. It might be different in a community maintained version of manim that is used by many people in different situations.

See https://github.com/3b1b/manim/pull/745 for discussion

PhilippImhof commented:
This PR has been sent to 3b1b's repository a few months ago, but never got merged as (probably) Grant did not consider this useful to his own projects. It might be different in a community maintained version of manim that is used by many people in different situations.

See https://github.com/3b1b/manim/pull/765 for original pull request.

leotrs commented:
LGTM. Thank you for your contribution!

leotrs commented:
BTW: in this repo, at least two members with write access have to leave an approving code review. I just left mine. We need one more so please be patient!

PhilippImhof commented:
Thank you Leo. I have been waiting in vain for quite a few months for the PR to get merged into 3b1b's manim repository, so I think I can easily wait for a second approval vote on this one :)

PhilippImhof commented:
Thank you @Aathish04. I will edit the PR text accordingly.

PgBiel commented on manimlib/animation/transform.py:
```diff
@@ -152,7 +152,7 @@ def __init__(self, method, *args, **kwargs):
         that method.  Key word arguments should be passed in
         as the last arg, as a dict, since **kwargs is for
-        configuration of the transform itslef
+        configuration of the transform itself
```
(As I stated elsewhere) Should probably be in a separate general cleanup PR (for organization purposes)

leotrs replied:
I'm fine with doing small typo fixes while in other PRs. If this contained dozens of doc/typo fixes then I'd ask for separate PRs.

PhilippImhof replied:
Actually, I do not know why this change is in my PR as it is not my commit. I pulled from master and created my branch for the matrix changes.

PgBiel commented on manimlib/mobject/matrix.py:
```diff
@@ -100,8 +102,8 @@ def organize_mob_matrix(self, matrix):
 
-    def add_brackets(self):
-        bracket_pair = TexMobject("\\big[", "\\big]")
+    def add_brackets(self, left, right):
```
Should probably default them to `r"\big[", r"\big]"`, jic (also would make this a non-breaking change)

leotrs replied:
Agreed

PhilippImhof replied:
Fine with my. Sorry for asking, but it seems to me that your suggested change does not use `r"\big["` and `r"\big]"` contrary to what you wrote in your comment. What do we do?

safinsingh replied:
That's because he used raw strings in his comment but non-raw strings in the edit; they do the same thing.

PhilippImhof replied:
@safinsingh Thanks, so I misunderstood. I thought his suggestion was mainly about using raw strings :)

PgBiel replied:
yeah sorry 
I prefer raw strings but I wanted to be consistent with the already-existing code

PgBiel commented on manimlib/mobject/svg/brace.py:
```diff
@@ -9,7 +9,7 @@
 from manimlib.utils.config_ops import digest_config
 from manimlib.utils.space_ops import get_norm
-
+import copy
```
(As I stated elsewhere) Should probably be in a separate general cleanup PR (for organization purposes)

PhilippImhof replied:
Same as above: This change should not be in my PR, as it has nothing to do with my changes.

PgBiel replied:
@PhillipImhof I added a suggestion here as well

PgBiel commented:
Was this tested?

PgBiel commented on manimlib/mobject/matrix.py:
```diff
@@ -125,6 +127,18 @@ def set_column_colors(self, *colors):
             column.set_color(color)
         return self
 
+    def get_rows(self):
```
Documentation

PgBiel commented on manimlib/mobject/matrix.py:
```diff
@@ -125,6 +127,18 @@ def set_column_colors(self, *colors):
+            for i in range(self.mob_matrix.shape[1])
+        ])
+
+    def set_row_colors(self, *colors):
```
Documentation

PhilippImhof commented:
@PgBiel Yes, of course, the code was tested. I have used it in several of my own animations.

I will add documentation as requested. Please understand that when I first wrote this, it was meant for 3b1b's repo and therefore follows Grant's verbosity....

PgBiel commented:
> @PgBiel Yes, of course, the code was tested. I have used it in several of my own animations.
Awesome! We should add Travis tests in the future as well

> I will add documentation as requested. Please understand that when I first wrote this, it was meant for 3b1b's repo and therefore follows Grant's verbosity....
Alright fam, I'll be waiting!

PhilippImhof commented:
I added short documentation and (hopefully) removed rogue commits by reverting them. However, I do not quite understand why the branch now has "conflicts that must be resolved" whereas there were no conflicts yesterday.

PhilippImhof commented:
@leotrs I did a `git rebase -i origin/master` and removed all rogue commits. I suppose the problem comes from (me misusing) GitHub Desktop. Instead of just pushing my changes, it magically pulled other commits and added them to the list. One never stops learning.

leotrs commented:
@PhilippImhof the conflicts are because there's a ton of activity right now on the repo and other PRs have been berged before yours. Not your fault :)

leotrs commented:
Now we have to wait for @PgBiel's pending review

PhilippImhof commented:
@leotrs Thanks. I somehow thought that GitHub would keep track of these changes, but in hindsight, that seems to be a stupid misconception as it would quickly become way too complicated.

leotrs commented:
@PhilippImhof and/or @PgBiel would you be so kind as to resolve the conflicts and then merge this? Thanks :)

PhilippImhof commented:
@leotrs I would gladly do so, but I do not have the necessary permissions.

PgBiel commented:
> @PhilippImhof and/or @PgBiel would you be so kind as to resolve the conflicts and then merge this? Thanks :)

I'll take a look in a bit
